Project/DAQ_System/Program/Src/Sensor/Accelerometer/LSM6DSOXAccelerometer.cpp

```diff
@@ -0,0 +1,118 @@
+    HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, commands,kNumBytes, HAL_MAX_DELAY);
+
+    calibrate();
+}
+
+void LSM6DSOX::calibrate() {
+    static constexpr uint8_t kSampleCount = 100;
+    double acc_magnitude = 0;
+    double average_array[kSampleCount];
+
+    // Collects samples to calibrate for 1G
+    for(int i = 0; i < kSampleCount; i++) {
+        ReadRawAcceleration();
+        acc_magnitude = sqrt(raw_acceleration_data_[0] * raw_acceleration_data_[0] + raw_acceleration_data_[1] * raw_acceleration_data_[1] + raw_acceleration_data_[2] * raw_acceleration_data_[2]);
+        average_array[i] = acc_magnitude * 0.001;
+    }
+
+    double min_value = std::numeric_limits<double>::max();
+    double max_value = -1;
+
+    // Find the largest and smallest magnitudes
+    for(int j = 2; j < kSampleCount; j++) {
+        if(average_array[j] <= min_value) {
+            min_value = average_array[j];
+
+        } else if (average_array[j] > max_value) {
+            max_value = average_array[j];
+        }
+    }
+
+    // Check that the maximum deviation is below a certain threshold.
+    // If it is too large, the conversion factor will be set to a previously
+    // determined conversion factor
+    gravity_adjustment_conversion_factor_ = 0;
+
+    if (max_value - min_value < 0.09) {
+        double total_array = 0;
+	    for(int i = 0; i < kSampleCount; i++) {
+		    total_array += average_array[i];
+	    }
+
+	    gravity_adjustment_conversion_factor_ = total_array / kSampleCount;
+    } else {
+        // Testing found that ~1.028 is about 1G
+        gravity_adjustment_conversion_factor_ = 1.028;
+    }
+}
+
+float* LSM6DSOX::GetAcceleration() {
+    return real_acceleration_data_;
+}
+
+void LSM6DSOX::ComputeAcceleration() {
+    ReadRawAcceleration();
+/*   old calculations for accelerometer
+    static constexpr float kMG_LSB = .001F;
+    real_acceleration_data_[0] = (raw_acceleration_data_[0]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[1] = (raw_acceleration_data_[1]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[2] = (raw_acceleration_data_[2]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+*/
+
+    // linearization of the accelerometer. by default range of the accelerometer is ±2 g
+    real_acceleration_data_[0] = 0.0006 *raw_acceleration_data_[0] - 0.0002;
+    real_acceleration_data_[1] = 0.0006 *raw_acceleration_data_[1] - 0.0002;
+    real_acceleration_data_[2] = 0.0006 *raw_acceleration_data_[2] - 0.0002;
+}
+
+void LSM6DSOX::ReadRawAcceleration() {
+	static constexpr int ByteArraySize = 6;
+	static constexpr int I2CWriteSize = 1;
+	static constexpr int I2CReadSize = 6;
+
+    uint8_t command[1] = { OUT_X_L_A };
+    HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, command,I2CWriteSize, HAL_MAX_DELAY);
+
+    uint8_t bytes_received[ByteArraySize];
+    HAL_I2C_Master_Receive(&i2c_,ACC_ADDRESS, bytes_received,I2CReadSize, HAL_MAX_DELAY);
```

Comment:
Shouldn't the `HAL_I2C_Master_Receive()` require the LSB of `ACC_ADDRESS` to be 1? That is, a hex value of `0xD5` instead of `0xD4`. Reference: [Table 11 of Page 37](https://www.mouser.com/datasheet/2/389/lsm6dsox-1517419.pdf).

Reply:
I'm not sure what part of the table you are referring. If you are talking about the SAD[0] bit then the answer is both yes and no. SAD[0] is connected to a physical pin on the board called DO that allows you to toggle the LSB of the slave address.